GVFS/GVFS.Platform.Windows/WindowsFileSystem.cs

```diff
@@ -117,6 +117,58 @@ public bool IsSocket(string fileName)
             return false;
         }
 
+        public void CreateDirectoryAccessibleByAuthUsers(string directoryPath)
+        {
+            if (Directory.Exists(directoryPath))
+            {
+                return;
+            }
+
+            // Create the ancestors of the specified path with the default ACLs
+            string parentPath = Path.GetDirectoryName(directoryPath);
+            Directory.CreateDirectory(parentPath);
+
+            // Create a temporary directory and read its ACLs.  This allows us to call
+            // Directory.CreateDirectorywith both the proper inherited ACLs, and the ACLs we want to add.
+            string tempDir = Path.Combine(parentPath, $"gvfs_{Path.GetRandomFileName()}");
+            Directory.CreateDirectory(tempDir);
+            DirectorySecurity directorySecurity;
+            try
+            {
+                directorySecurity = Directory.GetAccessControl(tempDir);
+            }
+            finally
+            {
+                Directory.Delete(tempDir);
+            }
+
+            // The following permissions are typically present on deskop and missing on Server
+            //
+            //   ACCESS_ALLOWED_ACE_TYPE: NT AUTHORITY\Authenticated Users
+            //          [OBJECT_INHERIT_ACE]
+            //          [CONTAINER_INHERIT_ACE]
+            //          [INHERIT_ONLY_ACE]
+            //        DELETE
+            //        GENERIC_EXECUTE
+            //        GENERIC_WRITE
+            //        GENERIC_READ
+
+            // Use AccessRuleFactory rather than creating a FileSystemAccessRule because the NativeMethods.FileAccess flags
+            // we're specifying are not valid for the FileSystemRights parameter of the FileSystemAccessRule constructor
+            AccessRule authenticatedUsersAccessRule = directorySecurity.AccessRuleFactory(
+                new SecurityIdentifier(WellKnownSidType.AuthenticatedUserSid, null),
+                unchecked((int)(NativeMethods.FileAccess.DELETE | NativeMethods.FileAccess.GENERIC_EXECUTE | NativeMethods.FileAccess.GENERIC_WRITE | NativeMethods.FileAccess.GENERIC_READ)),
+                true,
+                InheritanceFlags.ContainerInherit | InheritanceFlags.ObjectInherit,
+                PropagationFlags.None,
+                AccessControlType.Allow);
+
+            // The return type of the AccessRuleFactory method is the base class, AccessRule, but the return value can be cast safely to the derived class.
+            // https://msdn.microsoft.com/en-us/library/system.security.accesscontrol.filesystemsecurity.accessrulefactory(v=vs.110).aspx
+            directorySecurity.AddAccessRule((FileSystemAccessRule)authenticatedUsersAccessRule);
+            Directory.CreateDirectory(directoryPath, directorySecurity);

```

Comment:
Just a warning that this won't port easily to any .NET Core 3 codebases...

Can we not just create the final/intended directory, then _add_ the desired ACLs to that `DirectoryInfo` instance, rather than doing the temp dir create first?

Reply:
> Just a warning that this won't port easily to any .NET Core 3 codebases...

I saw that when looking through the documentation ☹️  

> Can we not just create the final/intended directory, then add the desired ACLs to that DirectoryInfo instance, rather than doing the temp dir create first?

That is definitely an option, and I think for the purposes of this PR it would work just as well (which makes me think perhaps for now I should switch to that approach 😄)

I went with the current approach with #1579 in mind.  My concern with create+adjust had been a race between two processes being interleaved like this:

1. Process 1 (admin): CreateDirectory
2. Process 2 (non-admin) CreateDirectory (no-op)
3. Process 2 (non-admin) SetACLs (_might_ fail as they are not the creator)
4. Process 2 (non-admin) Try to do stuff in the directory that requires the correct ACLs
5. Process 1 (admin) SetACLs (succeeds)

However, after giving it more thought over the weekend I might be overcomplicating that scenario, as I'm not sure that Process 2 would immediately perform operations that need the adjusted ACLs (and if the operation is something like downloading a loose object, it should use a retry+backoff if it fails).

I will update the code to your suggestion:

> just create the final/intended directory, then add the desired ACLs to that DirectoryInfo instance

And we can revisit the decision in #1579 if it ends up being an issue.